src/java/org/apache/cassandra/service/accord/CommandsForRanges.java

```diff
@@ -0,0 +1,505 @@
+            return (T) this;
+        }
+
+        private void addRanges(RangeCommandSummary summary)
+        {
+            for (Range range : summary.ranges)
+            {
+                rangeToTxn.add(Interval.create(normalize(range.start(), range.startInclusive(), true),
+                                               normalize(range.end(), range.endInclusive(), false),
+                                               summary));
+            }
+        }
+
+        public T putAll(CommandsForRanges other)
+        {
+            for (TxnId id : other.localCommands)
+            {
+                TxnType thisType = type(id);
+                TxnType otherType = other.type(id);
+                Invariants.checkArgument(thisType.isSafeToMix(otherType), "Attempted to add %s; expected %s but was %s", id, thisType, otherType);
+            }
+            localTxns.addAll(other.localCommands);
+            txnToRange.putAll(other.commandsToRanges);
+            // If "put" was called before for a txn present in "other", to respect the "put" semantics that update must
+            // be removed from "rangeToTxn" (as it got removed from "txnToRange").
+            // The expected common case is that this method is called on an empty builder, so the removeIf is off an
+            // empty list (aka no-op)
+            rangeToTxn.removeIf(data -> other.commandsToRanges.containsKey(data.txnId));
+            rangeToTxn.addAll(other.rangesToCommands);
+            return (T) this;
+        }
+
+        public T mergeRemote(TxnId txnId, Ranges ranges, BiFunction<? super Ranges, ? super Ranges, ? extends Ranges> remappingFunction)
+        {
+            Invariants.checkArgument(!localTxns.contains(txnId), "Attempted to merge remote txn %s, but this is a local txn", txnId);
+            // accord.impl.CommandTimeseries.mapReduce does the check on status and deps type, and NotWitnessed should match the semantics hard coded in InMemorySafeStore...
+            // in that store, the remote history is only ever included when minStauts == null and deps == ANY... but mapReduce sees accord.local.Status.KnownDeps.hasProposedOrDecidedDeps == false
+            // as a mis-match, so will be excluded... since NotWitnessed will return false it will only be included IFF deps = ANY.
+            // When it comes to the minStatus check, the current usage is "null", "Committed", "Accepted"... so NotWitnessed will only be included in the null case;
+            // the only subtle difference is if minStatus = NotWitnessed, this API will include these but InMemoryStore won't
+            RangeCommandSummary oldValue = txnToRange.get(txnId);
+            RangeCommandSummary newValue = oldValue == null ?
+                                           new RangeCommandSummary(txnId, ranges, SaveStatus.NotWitnessed, null, Collections.emptyList())
+                                           : oldValue.withRanges(ranges, remappingFunction);
+            if (newValue == null)
```

Comment:
Can `newValue` actually be null here? `oldValue.withRanges` doesn't appear to return null.

Reply:
in the current caller, no, but from a semantic point of view yes (as we match `Map.merge`).